src/data/upsert.ts

```diff
@@ -14,34 +11,81 @@ export const SparseValuesSchema = Type.Object(
   { additionalProperties: false }
 );
 
-const Vector = Type.Object(
+const VectorSchema = Type.Object(
   {
-    id: nonemptyString,
+    id: Type.String({ minLength: 1 }),
     values: Type.Array(Type.Number()),
     sparseValues: Type.Optional(SparseValuesSchema),
     metadata: Type.Optional(Type.Object({}, { additionalProperties: true })),
   },
   { additionalProperties: false }
 );
+const VectorArraySchema = Type.Array(VectorSchema);
 
-const VectorArray = Type.Array(Vector);
+const ChunkedUpsertSchema = Type.Object({
+  vectors: VectorArraySchema,
+  chunkSize: Type.Number({ minimum: 1 }),
+});
 
-export type Vector = Static<typeof Vector>;
+const UpsertOptionsSchema = Type.Union([
+  VectorArraySchema,
+  ChunkedUpsertSchema,
+]);
+
+export type Vector = Static<typeof VectorSchema>;
 export type SparseValues = Static<typeof SparseValuesSchema>;
-export type VectorArray = Static<typeof VectorArray>;
+export type VectorArray = Static<typeof VectorArraySchema>;
+export type UpsertOptions = Static<typeof UpsertOptionsSchema>;
 
 export const upsert = (api: VectorOperationsApi, namespace: string) => {
-  const validator = buildConfigValidator(VectorArray, 'upsert');
+  const validator = buildConfigValidator(UpsertOptionsSchema, 'upsert');
+
+  return async (options: UpsertOptions): Promise<void> => {
+    validator(options);
 
-  return async (vectors: VectorArray): Promise<void> => {
-    validator(vectors);
+    const isChunkedUpsert = !Array.isArray(options);
 
     try {
-      await api.upsert({ upsertRequest: { vectors, namespace } });
+      if (isChunkedUpsert) {
+        await chunkedUpsert(api, options.vectors, namespace, options.chunkSize);
+      } else {
+        await api.upsert({ upsertRequest: { vectors: options, namespace } });
+      }
       return;
     } catch (e) {
       const err = await handleDataError(e);
       throw err;
     }
   };
 };
+
+const chunkedUpsert = async (
+  api: VectorOperationsApi,
+  vectors: VectorArray,
+  namespace: string,
+  chunkSize = 10
+) => {
+  // Split vectors into chunks and promises for upserting
+  const chunks = sliceArrayToChunks(vectors, chunkSize);
+
+  const chunkedPromises = chunks.map(async (chunk) => {
+    try {
+      await api.upsert({ upsertRequest: { vectors: chunk, namespace } });
+    } catch (e) {
+      throw new Error(`Error upserting chunk: ${e}`);
+    }
+  });
+
+  try {
+    await Promise.allSettled(chunkedPromises);
+    return;
+  } catch (e) {
+    throw new Error(`Error upserting vectors into index: ${e}`);
+  }
+};
+
+const sliceArrayToChunks = (array: VectorArray, chunkSize: number) => {
```

Comment:
Since this code is dense and mathy, I think it's worth exporting this function (locally only, not from the published module) so you can test the chunk logic standalone. I'd want to verify we're not accidentally off-by-1 for various cases and going to accidentally leave behind a vector that should've been upserted (or issue include a vector in multiple upserts). By testing it separately, you can test with simpler arrays without the need to setup entire vector objects.

Potential cases to cover:
- array length 0
- array length 1
- array length even
- array length odd
- array length < chunkSize
- array length > chunkSize
- array divides evenly into chunkSize pieces
- array does not divide evenly into chunkSize pieces
- etc

Reply:
I believe I've covered all these cases along with `batchSize = 0`, which shouldn't occur via our code but figured I'd account for it just to be safe.

Great callout, let me know if there are other scenarios that should be expressed via unit tests.